src/seals/diagnostics/sort.py

```diff
@@ -0,0 +1,58 @@
+"""Environment to sort a list using swap actions."""
+
+from gym import spaces
+import numpy as np
+
+from seals import base_envs
+
+
+class SortEnv(base_envs.ResettableMDP):
+    """Environment to sort a list using swap actions."""
+
+    def __init__(self, length: int = 4):
+        """Constructs environment.
+
+        The initial state is a vector x sampled uniformly from
+        [0,1]**L, with actions a = (i,j) swapping x_i and x_j.  The
+        reward is given according to the number of elements in the
+        correct position.  To perform well, the learned policy must
+        compare elements, otherwise it will not generalize to all
+        possible randomly selected initial states.
+        """
+        self._length = length
+
+        super().__init__(
+            state_space=spaces.Box(low=0, high=1.0, shape=(length,)),
+            action_space=spaces.MultiDiscrete([length, length]),
+        )
+
+    def terminal(self, state: np.ndarray, n_actions_taken: int) -> bool:
+        """Always returns False."""
+        return False
+
+    def initial_state(self):
+        """Sample random vector uniformly in [0, 1]**L."""
+        return self.rand_state.random(size=self._length)
+
+    def reward(
+        self, state: np.ndarray, action: np.ndarray, new_state: np.ndarray,
+    ) -> float:
+        """Rewards fully sorted lists, and new correct positions."""
+        num_correct = self._num_correct_positions(state)
```

Comment:
Worth adding a comment noting that this isn't strictly potential shaping in the finite-horizon case since we do not return to a fixed-potential state at episode termination. (I don't think we are trying to be potential shaping here so I don't think we need to change the implementation, just avoid confusion.)